docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,320 @@
+// be applied to a container.
+type SecurityContextConstraints struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// AllowPrivilegedContainer determines if a container can request to be run as privileged.
+	AllowPrivilegedContainer bool `json:"allowPrivilegedContainer,omitempty" description:"allow containers to run as privileged"`
+	// AllowedCapabilities is a list of capabilities that can be requested to add to the container.
+	AllowedCapabilities []CapabilityType `json:"allowedCapabilities,omitempty" description:"capabilities that are allowed to be added"`
+	// AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
+	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin,omitempty" description:"allow the use of the host dir volume plugin"`
+	// SELinuxContext is the strategy that will dictate what labels will be set in the SecurityContext.
+	SELinuxContext SELinuxContextStrategyOptions `json:"seLinuxContext,omitempty" description:"strategy used to generate SELinuxOptions"`
+	// RunAsUser is the strategy that will dictate what RunAsUser is used in the SecurityContext.
+	RunAsUser RunAsUserStrategyOptions `json:"runAsUser,omitempty" description:"strategy used to generate RunAsUser"`
+
+	// The users who have permissions to use this security context constraints
+	Users []string `json:"users,omitempty" description:"users allowed to use this SecurityContextConstraints"`
+	// The groups that have permission to use this security context constraints
+	Groups []string `json:"groups,omitempty" description:"groups allowed to use this SecurityContextConstraints"`
+}
+
+// SELinuxContextStrategyOptions defines the strategy type and any options used to create the strategy.
+type SELinuxContextStrategyOptions struct {
+	// Type is the strategy that will dictate what SELinux context is used in the SecurityContext.
+	Type SELinuxContextStrategyType `json:"type,omitempty" description:"strategy used to generate the SELinux context"`
+
+	//TODO
+}
+
+// RunAsUserStrategyOptions defines the strategy type and any options used to create the strategy.
+type RunAsUserStrategyOptions struct {
+	// Type is the strategy that will dictate what RunAsUser is used in the SecurityContext.
+	Type RunAsUserStrategyType `json:"type,omitempty" description:"strategy used to generate RunAsUser"`
+	// UID is the user id that containers must run as.  Required for the MustRunAs strategy if not using
+	// namespace/service account allocated uids.
+	UID *int64 `json:"uid,omitempty" description:"the uid to always run as; required for MustRunAs"`
+	// AllocatedIDAnnotation provides an annotation that the strategy can look for if it should
+	// use a pre-allocated UID that exists on the namespace or service account.
+	AllocatedIDAnnotation string `json:"allocatedIDAnnotation,omitempty" description:"how the strategy can discover pre-allocated ids on the namespace or service account"`
```

Comment:
@erictune - is the red flag with the field that is is named `AllocatedIDAnnotation` or that is is actually storing what would be used to look up an annotation value at a later time?  

@smarterclayton - So my thoughts for this factoring was based on responsibilities broken down by:
Admission

1.  Choosing which SCC to use
1.  Calling an SCC provider to generate the SC based on the SCC
1.  Calling an SCC provider to validate the generated SC against the SCC
1.  Providing an accept/reject based on the above

SCC provider's

1.  Generate an SC from an SCC (where strategy based fields on the SCC can be used for delegation)
1.  Validate an SC against an SCC (where strategy based fields on the SCC can be used for delegation)

If the generate part of the SCC provider moves into the Admission controller it's the same complexity, the admission controller must still know a) when to try and resolve a UID based on the SCC being used (ie. RunAsAny vs MustRunAs) and b) how to look it up (what annotation to check and checking ns vs sa).